### PR TITLE
Store new refresh token in the state.json

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/COM-509

Changes:
- New `refresh_token` is stored to `state.json`, so it doesn't expires after 90 days.

### CODE REVIEW

File: src/Component.php [hunk 1]
```diff
@@ -4,29 +4,44 @@
...
     private Extractor $extractor;
 
+    private ArrayObject $stateObject;
+
     public function __construct(LoggerInterface $logger)
     {
         parent::__construct($logger);
         $config = $this->getConfig();
-        $accessTokenFactory = new AccessTokenFactory(
+        $this->stateObject = new ArrayObject($this->getInputState());
```
Comment (michaljurecko) [inline src/Component.php:23]: Zo `state` spravim object, aby som ho mohol citat a aj menit.

File: src/Component.php [hunk 2]
```diff
@@ -4,29 +4,44 @@
...
+        $clientFactory = new ClientFactory($tokenProvider, $config->getSubscriptionId());
         $this->extractor = new Extractor($this->getLogger(), $clientFactory->create());
     }
 
+    public function execute(): void
+    {
+        try {
+            parent::execute();
+        } finally {
+            $this->writeOutputStateToFile($this->stateObject->getArrayCopy());
```
Comment (michaljurecko) [inline src/Component.php:39]: Po skonceni behu komponenty sa zapise aktualizovany `state.json`.

File: src/OAuth/TokenDataManager.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                sprintf('Missing key "%s" in OAuth data array.', implode('", "', $missingKeys))
+            );
+        }
+    }
+
+    public function load(): array
+    {
+        // Load tokens from state.json
+        $authDataJson = $this->state[self::STATE_AUTH_DATA_KEY] ?? null;
+        if (is_string($authDataJson)) {
```
Comment (michaljurecko) [inline src/OAuth/TokenDataManager.php:38]: A zo `state.json` nacitam tokeny, ak su pritomne (teda novsie ako v konfiguracii).

File: src/OAuth/TokenDataManager.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            return JsonHelper::decode($authDataJson);
+        }
+
+        // Or use default from the configuration
+        return $this->configAuthData;
```
Comment (michaljurecko) [inline src/OAuth/TokenDataManager.php:43]: Ak nie su, tak pouzijem tie z konfiguracie.

File: src/OAuth/TokenDataManager.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+namespace Keboola\AzureCostExtractor\OAuth;
+
+use ArrayObject;
+use Keboola\Component\JsonHelper;
+use League\OAuth2\Client\Token\AccessTokenInterface;
+use Keboola\AzureCostExtractor\Exception\AccessTokenInitException;
+
+class TokenDataManager
+{
+    public const STATE_AUTH_DATA_KEY = '#refreshed_auth_data'; // # -> must be encrypted!
```
Comment (michaljurecko) [inline src/OAuth/TokenDataManager.php:14]: `state.json` automaticky hodnoty sifruje ak zacinaju `#`: https://developers.keboola.com/extend/common-interface/config-file/#state-file

File: tests/phpunit/OAuthTest.php [hunk 6]
```diff
@@ -4,18 +4,106 @@
...
+        $dataRaw = $state[TokenDataManager::STATE_AUTH_DATA_KEY];
+        Assert::assertIsString($dataRaw);
+        $data = JsonHelper::decode((string) $dataRaw);
+        Assert::assertNotEmpty($data['access_token']);
+        Assert::assertNotEmpty($data['refresh_token']);
+        Assert::assertNotSame($originAccessToken, $data['access_token']);
+        Assert::assertNotSame($originRefreshToken, $data['refresh_token']);
+    }
+
+    public function testStateInvalidTokens(): void
```
Comment (michaljurecko) [inline tests/phpunit/OAuthTest.php:90]: Nevalidne tokeny v `state.json`.

File: tests/phpunit/OAuthTest.php [hunk 7]
```diff
@@ -4,18 +4,106 @@
...
+use ArrayObject;
+use Keboola\AzureCostExtractor\Exception\AccessTokenRefreshException;
+use Keboola\AzureCostExtractor\OAuth\TokenDataManager;
+use Keboola\Component\JsonHelper;
 use PHPUnit\Framework\Assert;
 
 class OAuthTest extends BaseTest
 {
-    public function testRefreshToken(): void
+    public function testEmptyState(): void
```
Comment (michaljurecko) [inline tests/phpunit/OAuthTest.php:15]: Test: prazdny state.json, pouziju sa tokeny z configu

File: tests/phpunit/OAuthTest.php [hunk 8]
```diff
@@ -4,18 +4,106 @@
...
+        $state = $this->state->getArrayCopy();
+        $dataRaw = $state[TokenDataManager::STATE_AUTH_DATA_KEY];
+        $data = JsonHelper::decode($dataRaw);
+        Assert::assertNotEmpty($data['access_token']);
+        Assert::assertNotEmpty($data['refresh_token']);
+        Assert::assertNotSame($originAccessToken, $data['access_token']);
+        Assert::assertNotSame($originRefreshToken, $data['refresh_token']);
+    }
+
+    public function testEmptyStateInvalidTokens(): void
```
Comment (michaljurecko) [inline tests/phpunit/OAuthTest.php:41]: Test: prazdny state.json, nevalidne tokeny v configu.

File: tests/phpunit/OAuthTest.php [hunk 9]
```diff
@@ -4,18 +4,106 @@
...
+        ]);
+
+        $this->expectException(AccessTokenRefreshException::class);
+        $this->expectExceptionMessage(
+            'Microsoft OAuth API token refresh failed, please reset authorization in the extractor configuration.'
+        );
+        $tokenProvider->get();
+    }
+
+    public function testState(): void
```
Comment (michaljurecko) [inline tests/phpunit/OAuthTest.php:55]: Test: tokeny v state.json

**Review by ondrajodas — APPROVED** [review]
ok

**Pull request merged**